mem: `mlockall` pages to prevent swap bypassing storage fault tolerance (#2478)
tigerbeetle/tigerbeetle: docs/operating/linux.md

The service was created assuming it'll be used in a production scenario.

In case you want to use this service for development as well, you may need to adjust the `ExecStart` line to include the `--development` flag if your development environment doesn't support Direct IO, or if you require smaller cache sizes and/or batch sizes due to memory constraints.

### Memory Locking

> TigerBeetle relies on `RLIMIT_MEMLOCK` being adequate enough to initialize io_uring. It also locks all allocated memory to avoid the kernel swapping pages to disk and bypassing TigerBeetle's storage determinism. If enough memory cannot be locked, the environment can be appropriately modified either by disabling swap, adding the `CAP_IPC_LOCK` capability (provided by the systemd service), or raising the `memlock` value under `/etc/security/limits.conf`.

Reviewer: ```suggestion
TigerBeetle requires `RLIMIT_MEMLOCK` to be set high enough to:

1. initialize io_uring, which requires memory shared with the kernel to be locked, as well as
2. lock all allocated memory, and so prevent the kernel from swapping any pages to disk, which would not only affect performance but also bypass TigerBeetle's storage fault-tolerance.

If the required memory cannot be locked, then the environment should be modified either by (in order of preference):

1. giving the local `tigerbeetle` binary the `CAP_IPC_LOCK` capability, or
2. raising the global `memlock` value under `/etc/security/limits.conf`, or else
3. disabling swap (io_uring may still require an RLIMIT increase).
```

It's not about storage determinism, but storage fault-tolerance. Let's use bullets to make these reasons and mitigations clearer. And let's rank the mitigations in the best order (disabling swap is more a last resort than a first resort?).

> Memory locking is disabled for development environments using the `--development` flag.

Reviewer: ```suggestion
Memory locking is disabled for development environments when using the `--development` flag.
```

> For linux running under Docker, refer to [Allowing MEMLOCK](docker.md#allowing-memlock).

Reviewer: ```suggestion
For Linux running under Docker, refer to [Allowing MEMLOCK](docker.md#allowing-memlock).
```

Linux is a proper noun.

The service was created assuming it'll be used in a production scenario.

In case you want to use this service for development as well, you may need to adjust the `ExecStart` line to include the `--development` flag if your development environment doesn't support Direct IO, or if you require smaller cache sizes and/or batch sizes due to memory constraints.

### Memory Locking

TigerBeetle requires `RLIMIT_MEMLOCK` to be set high enough to:

1. initialize io_uring, which requires memory shared with the kernel to be locked, as well as
2. lock all allocated memory, and so prevent the kernel from swapping any pages to disk, which would not only affect performance but also bypass TigerBeetle's storage fault-tolerance.

If the required memory cannot be locked, then the environment should be modified either by (in order of preference):

> 1. giving the local `tigerbeetle` binary the `CAP_IPC_LOCK` capability, or

Reviewer 2: We could suggest the exact command: `sudo setcap "cap_ipc_lock=+ep" tigerbeetle`

Reviewer 2: (additionally, we could suggest it in the log warning message directly too.)

Author: I think suggesting in docs is better as it also has the commands in Docker to raise rlimits too.
